scipy/signal/spectral.py

```diff
@@ -267,121 +267,498 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
-                xft = fftpack.rfft(x_dt*win, nfft)
-                if k == 0:
-                    Pxx[..., 0] = xft[..., 0]**2
-                    Pxx[..., 1:] = xft[..., 1::2]**2 + xft[..., 2::2]**2
-                else:
-                    Pxx *= k/(k+1.0)
-                    Pxx[..., 0] += xft[..., 0]**2 / (k+1)
-                    Pxx[..., 1:] += (xft[..., 1::2]**2 + xft[..., 2::2]**2) \
-                                  / (k+1.0)
-
-        Pxx[..., 1:-1] *= 2*scale
-        Pxx[..., (0,-1)] *= scale
-        f = np.arange(Pxx.shape[-1]) * (fs/nfft)
+    if isinstance(window, string_types) or type(window) is tuple:
+        win = get_window(window, nperseg)
     else:
-        for k, ind in enumerate(indices):
-            x_dt = detrend_func(x[..., ind:ind+nperseg])
-            xft = fftpack.fft(x_dt*win, nfft)
-            if k == 0:
-                Pxx = (xft * xft.conj()).real
-            else:
-                Pxx *= k/(k+1.0)
-                Pxx += (xft * xft.conj()).real / (k+1.0)
-        Pxx *= scale
-        f = fftpack.fftfreq(nfft, 1.0/fs)
+        win = np.asarray(window)
+        if len(win.shape) != 1:
+            raise ValueError('window must be 1-D')
+        if win.shape[0] != nperseg:
+            raise ValueError('window must have length of nperseg')
 
+    if np.result_type(win,np.complex64) != outdtype:
+        win = win.astype(outdtype)
+
+    if mode == 'psd':
+        if scaling == 'density':
+            scale = 1.0 / (fs * (win*win).sum())
+        elif scaling == 'spectrum':
+            scale = 1.0 / win.sum()**2
+        else:
+            raise ValueError('Unknown scaling: %r' % scaling)
+    else:
+        scale = 1
+
+    if return_onesided is True:
+        if np.iscomplexobj(x):
+            sides = 'twosided'
+        else:
+            sides = 'onesided'
+            if not same_data:
+                if np.iscomplexobj(y):
+                    sides = 'twosided'
+    else:
+        sides = 'twosided'
+
+    if sides == 'twosided':
+        num_freqs = nfft
+    elif sides == 'onesided':
+        if nperseg % 2:
+            num_freqs = (nfft + 1)//2
+        else:
+            num_freqs = nfft//2 + 1
+
+    # Perform the windowed FFTs
+    result = _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft)
+    result = result[..., :num_freqs]
+    freqs = fftpack.fftfreq(nfft, 1/fs)[:num_freqs]
+
+    if not same_data:
+        # All the same operations on the y data
+        result_y = _fft_helper(y, win, detrend_func, nperseg, noverlap, nfft)
+        result_y = result_y[..., :num_freqs]
+        result = np.conjugate(result) * result_y
+    elif mode == 'psd':
+        result = np.conjugate(result) * result
+    elif mode == 'magnitude':
+        result = np.absolute(result)
+    elif mode == 'angle' or mode == 'phase':
+        result = np.angle(result)
+    elif mode == 'complex':
+        pass
+
+    result *= scale
+    if mode == 'psd' and sides == 'onesided':
```

Comment:
This line will need the same logic as was added in https://github.com/matplotlib/matplotlib/pull/4326